src/preamble.js

```diff
@@ -747,12 +747,16 @@ function instrumentWasmTableWithAbort() {
 #endif
 
 #if EXPORT_ES6
-// Use bundler-friendly `new URL(..., import.meta.url)` pattern; works in browsers too.
-var wasmBinaryFile = new URL('{{{ WASM_BINARY_FILE }}}', import.meta.url).toString();
-#else
-var wasmBinaryFile = '{{{ WASM_BINARY_FILE }}}';
-if (!isDataURI(wasmBinaryFile)) {
-  wasmBinaryFile = locateFile(wasmBinaryFile);
+if (Module['locateFile']) {
```

Comment:
checking for `Module['locateFile']` here feels a little odd, since we don't use it - we call `locateFile`, and the Module part is an internal detail there. but you do need to depend on that internal detail (whether the user provided a locateFile impl)...

I don't understand how this works, though. I thought the point of the `import.meta.url` pattern was to statically handle things in bundlers - but then how do things work if there is a locateFile? What does the bundler do?

Reply:
> checking for Module['locateFile'] here feels a little odd, since we don't use it - we call locateFile, and the Module part is an internal detail there. but you do need to depend on that internal detail (whether the user provided a locateFile impl)...

I agree. Ideally we could have probably deprecated `locateFile` for ES6 mode, but that ship has sailed plus it was the only way to locate files in bundlers before the `import.meta`-based integration. So now it's required to keep backward compatibility.

> What does the bundler do?

Bundler detects the `new URL(...)` pattern statically and replaces it with the final URL of the Wasm wherever it was put by the bundler. So code like above would be transformed in the end to:

```js
if (Module['locateFile']) {
  var wasmBinaryFile = 'some.wasm';
  if (!isDataURI(wasmBinaryFile)) {
    wasmBinaryFile = locateFile(wasmBinaryFile);
  }
} else {
  var wasmBinaryFile = '/path/to/wherever-bundler-put-the.wasm';
}
```

This way, bundler will always put Wasm into its final output directory and that path will be used by Emscripten by default, but if user defines `Module.locateFile` dynamically, it takes precedence instead.

Reply:
Interesting, thanks.

And there are cases where a user wants to both use a bundler to decide where to put the wasm, and also customize the location of the wasm themselves with locateFile? That's what I don't get. There is just one binary file here - is the idea that in one deployment the default bundler location would be used, and in another a custom place using locateFile? (if so, why is that needed? sorry, I don't know much about bundlers, so these may not be great questions...)

Reply:
> And there are cases where a user wants to both use a bundler to decide where to put the wasm, and also customize the location of the wasm themselves with locateFile? That's what I don't get.

In our case we ran into this because we didn't yet have Rollup plugin handling `new URL` (in Rollup pretty much everything is done via plugins), and before this integration our code used `locateFile` to specify the correct location, but that one got broken too, so Wasm couldn't be loaded via either approach.

It might be a somewhat niche usecase - e.g. in Webpack, which is more popular for apps, `new URL` is handled out of the box so we wouldn't notice the problem, but if others use Rollup they will likely run into this too hence the feeling of urgency.

Other than this unfortunate combination, you're right that `locateFile` shouldn't be necessary as Wasm should normally be alongside JS (if using ES6 output directly without a bundler), or the URL should be handled by a bundler, but... I don't know what people do in the wild. Maybe someone does use ES6 output directly yet moves Wasm to a custom location? Hard to tell until issue reports come in.

Reply:
>  Maybe someone does use ES6 output directly yet moves Wasm to a custom location? Hard to tell until issue reports come in.

I was actually thinking of later adding a diagnostic that would warn if `locateFile` returns same URL as computed via the added `new URL` - something along the lines "looks like you're using `locateFile` to specify location to the Wasm binary, but it's no longer needed in ES6 code". This could help catch legitimate use-cases if there are any, or deprecate that function in ES6 output otherwise.

Reply:
I think I see now, thanks!